calculator/calculator.py

```diff
@@ -1,15 +1,115 @@
+
+def _handle_parenthesis(token) -> str:
+    return str(_calculate(token.group("EXTRACTED_EXPRESSION")))
+
+def _handle_factorial(token) -> str:
+    return str(math.factorial(float(token.group("NUMBER"))))
+
+def _handle_negate(token) -> str:
+    return str(-1 * float(token.group("NUMBER")))
+
+def _handle_max_min_average(token) -> str:
+    if token.group("MAX"):
+        return str(max(float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))))
+    elif token.group("MIN"):
+        return str(min(float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))))
+    else:
+        return str(mean([float(token.group("FIRST_NUMBER")), float(token.group("SECOND_NUMBER"))]))
+
+def _handle_power(token) -> str:
+    return str(float(token.group("FIRST_NUMBER")) ** float(token.group("SECOND_NUMBER")))
+
+def _handle_multiplication_division(token) -> str:
+    if token.group("MULTIPLICATION"):
+        return str(float(token.group("FIRST_NUMBER")) * float(token.group("SECOND_NUMBER")))
+    else:
+        return str(float(token.group("FIRST_NUMBER")) / float(token.group("SECOND_NUMBER")))
+
+def _handle_addition_subtraction(token) -> str:
+    if token.group("ADDITION"):
+        return str(float(token.group("FIRST_NUMBER")) + float(token.group("SECOND_NUMBER")))
+    else:
+        return str(float(token.group("FIRST_NUMBER")) - float(token.group("SECOND_NUMBER")))
+
+
+_LEFT_PARENTHESIS_REGEX = r'(?P<LEFT_PARENTHESIS>\()'
+_RIGHT_PARENTHESIS_REGEX = r'(?P<RIGHT_PARENTHESIS>\))'
+_FACTORIAL_REGEX = r'(?P<FACTORIAL>!)'
+_NEGATE_REGEX = r'(?P<NEGATE>~)'
+_AVERAGE_REGEX = r'(?P<AVERAGE>@)'
+_MIN_REGEX = r'(?P<MIN>\&)'
+_MAX_REGEX = r'(?P<MAX>\$)'
+_POWER_REGEX = r'(?P<POWER>\^)'
+_MULTIPLICATION_REGEX = r'(?P<MULTIPLICATION>\*)'
+_DIVISION_REGEX = r'(?P<DIVISION>/)'
+_ADDITION_REGEX = r'(?P<ADDITION>\+)'
+_SUBTRACTION_REGEX = r'(?P<SUBTRACTION>-)'
+
+_TOKEN_SPECIFICATION = {
+    'PARENTHESIS_PAIR': (f'{_LEFT_PARENTHESIS_REGEX}'
+                         f'(?P<EXTRACTED_EXPRESSION>[^{_LEFT_PARENTHESIS_REGEX}{_RIGHT_PARENTHESIS_REGEX}]*)'
+                         f'{_RIGHT_PARENTHESIS_REGEX}', _handle_parenthesis),
+    'FACTORIAL': (_one_variable_operation(f'{_FACTORIAL_REGEX}', False), _handle_factorial),
+    'NEGATE': (_one_variable_operation(f'{_NEGATE_REGEX}', True), _handle_negate),
+    'AVERAGE_MAX_MIN': (_two_variable_operation(f'(?:{_AVERAGE_REGEX}|{_MIN_REGEX}|{_MAX_REGEX})', True),
+                        _handle_max_min_average),
+    'POWER': (_two_variable_operation(f'{_POWER_REGEX}', True), _handle_power),
+    'MULTIPLICATION_DIVISION': (_two_variable_operation(f'(?:{_MULTIPLICATION_REGEX}|{_DIVISION_REGEX})', True),
+                                _handle_multiplication_division),
+    'ADDITION_SUBTRACTION': (_two_variable_operation(f'(?:{_ADDITION_REGEX}|{_SUBTRACTION_REGEX})', True),
+                             _handle_addition_subtraction),
+}
```

Comment:
You specify for each regex an operation.
Think about splitting the regexes to atomic operations - regex for each single operation.
It will let you be more generic.